cfme/web_ui/expression_editor_widgetastic.py

```diff
@@ -0,0 +1,471 @@
+
+    def __locator__(self):
+        return self.ROOT
+
+    def click_undo(self):
+        self.browser.element(self.UNDO, self).click()
+
+    def click_redo(self):
+        self.browser.element(self.REDO, self).click()
+
+    def click_and(self):
+        self.browser.element(self.AND, self).click()
+
+    def click_or(self):
+        self.browser.element(self.OR, self).click()
+
+    def click_not(self):
+        self.browser.element(self.NOT, self).click()
+
+    def click_remove(self):
+        self.browser.element(self.REMOVE, self).click()
+
+    def click_commit(self):
+        self.browser.element(self.COMMIT, self).click()
+
+    def click_discard(self):
+        self.browser.element(self.DISCARD, self).click()
+
+    def click_switch_to_relative(self):
+        self.browser.element(self.SELECT_RELATIVE, self).click()
+
+    def click_switch_to_specific(self):
+        self.browser.element(self.SELECT_SPECIFIC, self).click()
+
+    @property
+    def _atom_root(self):
+        return self.browser.element(self.ATOM_ROOT, self)
+
+    @property
+    def _expressions_root(self):
+        return self.browser.element(self.EXPRESSIONS_ROOT, self)
+
+    def select_first_expression(self):
+        "There is always at least one (???), so no checking of bounds."
+        self.browser.elements("//a[contains(@id,'exp_')]", self._expressions_root)[0].click()
+
+    def select_expression_by_text(self, text):
+        self.browser.element(
+            "//a[contains(@id,'exp_')][contains(normalize-space(text()),'{}')]".format(text),
+            self
+        ).click()
+
+    def no_expression_present(self):
+        els = self.browser.elements("//a[contains(@id,'exp_')]", self._expressions_root)
+        if len(els) > 1:
+            return False
+        return els[0].text.strip() == "???"
+
+    def any_expression_present(self):
+        return not self.no_expression_present()
+
+    def is_editing(self):
+        try:
+            self.browser.element(
+                "//a[contains(@id,'exp_')][contains(normalize-space(text()),'???')]",
+                self._expressions_root
+            )
+            return True
+        except NoSuchElementException:
+            return False
+
+    def delete_whole_expression(self):
+        while self.any_expression_present():
+            self.select_first_expression()
+            self.click_remove()
+
+    def read(self):
+        "Returns whole expression as represented visually."
+        return self._expressions_root.text.encode("utf-8").strip()
+
+    def enable_editor(self):
+        try:
+            el = self.browser.element(self.show_loc, self)
+            wait_for(lambda: el.is_displayed, num_sec=2, delay=0.2)
+            el.click()
+        except (TimedOutError, NoSuchElementException):
+            pass
+
+    def fill(self, expression):
+        if self.show_loc is not None:
+            self.enable_editor()
+        prog = create_program(expression, self)
+        prog()
```

Comment:
fill() must return a boolean whether it changed anything (True) or not (False). I guess that if you read the expresion before and after and return != of them, that could be a good start.